From 4673a73e8558c984e8048f44e4f4b885e9878d13 Mon Sep 17 00:00:00 2001
From: Omar Ramirez Luna <omar.r.ramirez.luna@intel.com>
Date: Thu, 20 Oct 2016 14:06:01 -0700
Subject: [PATCH] ov8858: fix regulator management on ov8858_s_ctrl

ov8858_s_ctrl can also communicate through i2c with the sensor
if the voltages are not enabled it will cause a timeout in the
i2c line.

This will fix the bug uncovered by the usage of the regulator
framework:

ov8858 i2c-INT3477:00: gmin: initializing atomisp module subdev data.PMIC ID 1
ov8858 i2c-INT3477:00: camera pdata: port: 1 lanes: 4 order: 00000002
i2c_designware 808622C1:03: controller timed out
ov8858 i2c-INT3477:00: write DW9718_DATA_M  failed -5

Signed-off-by: Omar Ramirez Luna <omar.r.ramirez.luna@intel.com>
---
 drivers/media/i2c/ov8858.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/media/i2c/ov8858.c b/drivers/media/i2c/ov8858.c
index 8001995..9410504 100644
--- a/drivers/media/i2c/ov8858.c
+++ b/drivers/media/i2c/ov8858.c
@@ -1765,6 +1765,12 @@ static int ov8858_s_ctrl(struct v4l2_ctrl *ctrl)
 	 * doesn't need to be taken here.
 	 */
 
+	ret = __ov8858_s_power(&dev->sd, 1);
+	if (ret) {
+		dev_err(&client->dev, "power-up error %d!\n", ret);
+		return ret;
+	}
+
 	switch (ctrl->id) {
 	case V4L2_CID_RUN_MODE:
 		switch (ctrl->val) {
@@ -1809,6 +1815,8 @@ static int ov8858_s_ctrl(struct v4l2_ctrl *ctrl)
 		ret = -EINVAL;
 	}
 
+	__ov8858_s_power(&dev->sd, 0);
+
 	return ret;
 }
 
-- 
1.9.1

